controllers/topology/controller_test.go

```diff
@@ -0,0 +1,129 @@
+	*replicas = 3
+	md := clusterv1.MachineDeploymentTopology{
+		Class:    "linux-worker",
+		Replicas: replicas,
+		Name:     "md-linux-workers",
+	}
+	cluster1 := testtypes.NewClusterBuilder(ns.Name, "cluster1").
+		WithClusterClass(*cc).
+		WithTopologyVersion("1.22.1").
+		WithMachineDeploymentTopology([]clusterv1.MachineDeploymentTopology{md}).
+		Build()
+
+	tests := []struct {
+		name            string
+		startingCluster *clusterv1.Cluster
+		initObjs        []client.Object
+		want            *clusterv1.Cluster
+		wantErr         bool
+	}{
+		{
+			name: "Create cluster with ControlPlane, ControlPlaneMachineInfrastructure, MachineDeployments, and InfrastructureCluster",
+			initObjs: []client.Object{
+				cc,
+				infra,
+				cpInfra,
+				btstrp,
+				cp,
+			},
+			startingCluster: cluster1,
+			want: testtypes.NewClusterBuilder(ns.Name, "cluster1").
+				WithInfrastructureCluster(infrastructureCluster).
+				WithClusterClass(*cc).
+				WithLabels(map[string]string{"cluster.x-k8s.io/cluster-name": "cluster1", "topology.cluster.x-k8s.io/owned": ""}).
+				WithTopologyVersion("v1.22.1").
+				WithControlPlane(controlPlane).
+				WithMachineDeploymentTopology([]clusterv1.MachineDeploymentTopology{md}).
+				Build(),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+			// clusterComparer is the object that will be used to compare the resulting Cluster with the expected Cluster
+			clusterComparer := testtypes.ClusterComparers{
+				ObjectMeta:          testtypes.ObjectMetaComparer(tt.want).IgnoreByDefault().MatchNameNamespaceAnnotationsLabels(),
+				TypeMeta:            testtypes.TypeMetaComparer(tt.want),
+				ClusterSpecFields:   testtypes.ClusterSpecComparer(tt.want).CompatibleControlPlaneRef().CompatibleInfrastructureRef(),
+				ClusterStatusFields: testtypes.ClusterStatusComparer(tt.want).IgnoreByDefault(),
+			}
+
+			tt.initObjs = append(tt.initObjs, tt.startingCluster)
+			for _, obj := range tt.initObjs {
+				err := env.Create(ctx, obj)
+				g.Expect(err).ToNot(HaveOccurred())
+			}
+			key := client.ObjectKey{Name: "cluster1", Namespace: ns.Name}
+			got := &clusterv1.Cluster{}
+			g.Eventually(func() bool {
+				if err := env.Get(ctx, key, got); err != nil {
+					return false
+				}
+				// Initial test to see if InfrastructureCluster and ControlPlaneRef have been set in the cluster.
+				if got.Spec.InfrastructureRef == nil || got.Spec.ControlPlaneRef == nil {
+					return false
+				}
+				// Check to see if correct MachineDeployment has been created in the cluster.
+				md := &clusterv1.MachineDeploymentList{}
+				err := env.Client.List(ctx, md, client.MatchingLabels{
+					clusterv1.ClusterLabelName:          "cluster1",
+					clusterv1.ClusterTopologyOwnedLabel: "",
+				})
+				if err != nil {
+					return false
+				}
+
+				g.Expect(*got).To(clusterComparer.Matcher(), cmp.Diff(got, tt.want))
```

Comment:
This is a sort of unusual syntax for a matcher... I was expecting something like

```suggestion
				g.Expect(*got).To(clusterComparer.Match(tt.want), ...)
```

🤔

Reply:
Agreed